### PR TITLE
List of contents in the shared folder

### PR DESCRIPTION
**Pull request by Simon-Le**
- A list of the contents of a given folder gan be retrieved.
- Mocked SharedFile object now contains also the type of the file (file or directory).

### CODE REVIEW

File: pyndows/_windows.py [hunk 1]
```diff
@@ -125,6 +133,46 @@ def _rename(
...
 
+def get_folder_contents(
+    connection: SMBConnection,
+    share_folder: str,
+    path: str = "",
+    include_folders: bool = True,
+    pattern: str = "",
+) -> List[SharedFile]:
+    """
+    Returns a list of a folder contents which match a given pattern.
```
Comment (Colin-b) [inline pyndows/_windows.py]:
```suggestion
    Returns a list of files or folders matching given pattern within a folder.
```

File: pyndows/_windows.py [hunk 2]
```diff
@@ -125,6 +133,46 @@ def _rename(
         )
 
 
+def get_folder_contents(
+    connection: SMBConnection,
+    share_folder: str,
+    path: str = "",
```
Comment (Colin-b) [inline pyndows/_windows.py]: Can you rename to folder_path ?

File: pyndows/_windows.py [hunk 3]
```diff
@@ -125,6 +133,46 @@ def _rename(
...
+        SMB_FILE_ATTRIBUTE_READONLY
+        | SMB_FILE_ATTRIBUTE_HIDDEN
+        | SMB_FILE_ATTRIBUTE_SYSTEM
+        | SMB_FILE_ATTRIBUTE_ARCHIVE
+        | SMB_FILE_ATTRIBUTE_INCL_NORMAL
+    )
+    if include_folders:
+        search = search | SMB_FILE_ATTRIBUTE_DIRECTORY
+    logger.info(
+        f"Returning the list of contents at \\\\{connection.remote_name}\\{share_folder}\\{path} ..."
```
Comment (Colin-b) [inline pyndows/_windows.py]:
```suggestion
        f"Listing the content of \\\\{connection.remote_name}\\{share_folder}\\{path} ..."
```

File: pyndows/testing.py [hunk 4]
```diff
@@ -115,16 +121,47 @@ def retrieveFile(self, share_drive_path: str, file_path: str, file) -> (int, int
...
+        | SMB_FILE_ATTRIBUTE_INCL_NORMAL,
+        pattern: str = "*",
     ) -> List[SharedFile]:
-        files_list = [
-            SharedFileMock(os.path.basename(file_path))
-            for _, file_path in SMBConnectionMock.stored_files
-            if re.search(pattern, os.path.basename(file_path))
-        ]
+
+        pattern = "" if pattern == "*" else pattern
```
Comment (Colin-b) [inline pyndows/testing.py]: Doesn't it work ?
```suggestion
        pattern = pattern.replace("*", ".*")
```

File: pyndows/_windows.py [hunk 5]
```diff
@@ -125,6 +133,50 @@ def _rename(
...
+    """
+    search = (
+        SMB_FILE_ATTRIBUTE_READONLY
+        | SMB_FILE_ATTRIBUTE_ARCHIVE
+        | SMB_FILE_ATTRIBUTE_INCL_NORMAL
+    )
+    if include_folders:
+        search = search | SMB_FILE_ATTRIBUTE_DIRECTORY
+    logger.info(
+        f"Listing the content of \\\\{connection.remote_name}\\{share_folder}\\{path} ..."
```
Comment (Simon-Le) [inline pyndows/_windows.py]:
```suggestion
        f"Listing the content of \\\\{connection.remote_name}\\{share_folder}\\{folder_path} ..."
```

File: pyndows/testing.py [hunk 6]
```diff
@@ -115,16 +121,47 @@ def retrieveFile(self, share_drive_path: str, file_path: str, file) -> (int, int
...
+
+        def to_file_in_path(file_path: str, path: str) -> (str, bool):
```
Comment (Colin-b) [inline pyndows/testing.py]: Code should work with / or \ in paths as os.sep

File: CHANGELOG.md [hunk 7]
```diff
@@ -6,6 +6,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Added
+- Added get_folder_contents which returns list of the folders/files in the shared folder or one of its sub-folders.
```
Comment (Colin-b) [inline CHANGELOG.md]:
```suggestion
- pyndows.get_folder_content listing the content of a folder (non-recursively).
```

File: CHANGELOG.md [hunk 8]
```diff
@@ -6,6 +6,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+- Mocked SharedFile object now contains also the type of the file (file or directory).
```
Comment (Colin-b) [inline CHANGELOG.md]:
```suggestion
- File type (isDirectory boolean) within mocked SharedFile object.
```

File: pyndows/_windows.py [hunk 9]
```diff
@@ -1,9 +1,17 @@
...
 import logging
 import os
-from typing import Optional
-
-from smb.SMBConnection import SMBConnection
+from typing import Optional, List
+
+from smb.SMBConnection import (
+    SMBConnection,
+    SMB_FILE_ATTRIBUTE_READONLY,
```
Comment (Colin-b) [inline pyndows/_windows.py]: Please remove unused imports

File: pyndows/_windows.py [hunk 10]
```diff
@@ -125,6 +133,50 @@ def _rename(
...
 
+def get_folder_content(
+    connection: SMBConnection,
+    share_folder: str,
+    folder_path: str = "",
+    include_folders: bool = True,
+    pattern: str = "*",
+) -> List[SharedFile]:
+    """
+    Returns a list of files or folders matching given pattern within a folder.
```
Comment (Colin-b) [inline pyndows/_windows.py]: Explain that it is not recursive

File: pyndows/_windows.py [hunk 11]
```diff
@@ -125,6 +133,50 @@ def _rename(
...
+
+    :param connection: Samba connection.
+    :param share_folder: Remote computer name.
+    :param folder_path: The path of the folder for which the list of contents are returned.
```
Comment (Colin-b) [inline pyndows/_windows.py]:
```suggestion
    :param folder_path: Folder path for which the content is returned.
```

File: pyndows/_windows.py [hunk 12]
```diff
@@ -125,6 +133,50 @@ def _rename(
...
+    Should be a relative path from the share_folder.
```
Comment (Colin-b) [inline pyndows/_windows.py]:
```suggestion
    Must be relative to share_folder.
```

File: pyndows/_windows.py [hunk 13]
```diff
@@ -125,6 +133,50 @@ def _rename(
...
+    Defaults to an empty string which means the root of the shared folder
```
Comment (Colin-b) [inline pyndows/_windows.py]:
```suggestion
    Defaults to the root of the shared folder (empty string).
```

File: pyndows/_windows.py [hunk 14]
```diff
@@ -125,6 +133,50 @@ def _rename(
...
+    :param include_folders: Include folders in the returned list.
```
Comment (Colin-b) [inline pyndows/_windows.py]:
```suggestion
    :param include_folders: Should sub folders be included in the results.
```

File: pyndows/_windows.py [hunk 15]
```diff
@@ -125,6 +133,50 @@ def _rename(
...
+    Defaults to True, when it's set to false, only a list of files will be returned.
```
Comment (Colin-b) [inline pyndows/_windows.py]:
```suggestion
    Include sub folders by default. Set to False to list only files.
```

File: pyndows/_windows.py [hunk 16]
```diff
@@ -125,6 +133,50 @@ def _rename(
...
+    :param pattern: a regex pattern to match it on files/folders names.
```
Comment (Colin-b) [inline pyndows/_windows.py]:
```suggestion
    :param pattern: Filter out files or sub folders based on this pattern (`*` character means all).
```

File: pyndows/_windows.py [hunk 17]
```diff
@@ -125,6 +133,50 @@ def _rename(
...
+    Defaults to * which includes all the names.
```
Comment (Colin-b) [inline pyndows/_windows.py]:
```suggestion
    Include everything but . and .. by default (*).
```

File: pyndows/_windows.py [hunk 18]
```diff
@@ -125,6 +133,50 @@ def _rename(
...
+
```
Comment (Colin-b) [inline pyndows/_windows.py]:
```suggestion
```

File: pyndows/_windows.py [hunk 19]
```diff
@@ -125,6 +133,50 @@ def _rename(
...
+    :return: A List of SharedFile objects, an empty list if the given folder doesn't exist.
```
Comment (Colin-b) [inline pyndows/_windows.py]:
```suggestion
    :return: A List of SharedFile objects, empty if the given folder does not exist.
```

File: pyndows/_windows.py [hunk 20]
```diff
@@ -125,6 +133,50 @@ def _rename(
...
+    logger.info(
+        f"Listing the content of \\\\{connection.remote_name}\\{share_folder}\\{folder_path} ..."
+    )
+    try:
+        return [
+            file
+            for file in connection.listPath(
+                share_folder, folder_path, pattern=pattern, search=search
+            )
+            if file.filename != "." and file.filename != ".."
```
Comment (Colin-b) [inline pyndows/_windows.py]:
```suggestion
            if file.filename not in (".", "..")
```

File: pyndows/testing.py [hunk 21]
```diff
@@ -115,16 +121,47 @@ def retrieveFile(self, share_drive_path: str, file_path: str, file) -> (int, int
...
     def listPath(
-        self, service_name: str, path: str, pattern: str = "*"
+        self, service_name: str, path: str, search: int = 0, pattern: str = "*",
     ) -> List[SharedFile]:
-        files_list = [
-            SharedFileMock(os.path.basename(file_path))
-            for _, file_path in SMBConnectionMock.stored_files
-            if re.search(pattern, os.path.basename(file_path))
-        ]
+
```
Comment (Colin-b) [inline pyndows/testing.py]:
```suggestion
```

File: pyndows/testing.py [hunk 22]
```diff
@@ -115,16 +121,47 @@ def retrieveFile(self, share_drive_path: str, file_path: str, file) -> (int, int
...
+        def to_file_in_path(file_path: str, path: str) -> (str, bool):
+            file_path = file_path[len(path) :]
+            is_directory = os.sep in file_path
+            if is_directory:
+                return file_path.split(os.sep, maxsplit=1)[0], True
+            return file_path, False
+
+        path = (
+            service_name + path
+            if path.endswith(os.sep) or path == ""
```
Comment (Colin-b) [inline pyndows/testing.py]:
```suggestion
            if not path or path.endswith(os.sep)
```

File: pyndows/testing.py [hunk 23]
```diff
@@ -115,16 +121,40 @@ def retrieveFile(self, share_drive_path: str, file_path: str, file) -> (int, int
...
+        path = path.replace(os.altsep, os.sep)
+
+        def to_file_in_path(file_path: str, path: str) -> (str, bool):
+            file_path = file_path[len(path) :]
+            is_directory = os.sep in file_path
+            if is_directory:
+                return file_path.split(os.sep, maxsplit=1)[0], True
+            return file_path, False
+
+        path = path if not path or path.endswith(os.sep) else f"{path + os.sep}"
```
Comment (Colin-b) [inline pyndows/testing.py]:
```suggestion
        path = path if not path or path.endswith(os.sep) else f"{path}{os.sep}"
```

File: pyndows/testing.py [hunk 24]
```diff
@@ -115,16 +121,40 @@ def retrieveFile(self, share_drive_path: str, file_path: str, file) -> (int, int
...
+
+        files_list = []
+        for shared_folder, file_path in SMBConnectionMock.stored_files:
+            shared_folder = shared_folder.replace(os.altsep, os.sep)
+            file_path = file_path.replace(os.altsep, os.sep)
+            file_in_path = to_file_in_path(file_path, path)
```
Comment (Colin-b) [inline pyndows/testing.py]:
```suggestion
            file_name, is_directory = to_file_in_path(file_path, path)
```

File: pyndows/testing.py [hunk 25]
```diff
@@ -115,16 +121,40 @@ def retrieveFile(self, share_drive_path: str, file_path: str, file) -> (int, int
...
+            if (
+                shared_folder == service_name
+                and f"{os.path.dirname(file_path) + os.sep}".startswith(path)
+                and re.search(pattern, file_in_path[0])
```
Comment (Colin-b) [inline pyndows/testing.py]:
```suggestion
                and re.search(pattern, file_name)
```

File: pyndows/testing.py [hunk 26]
```diff
@@ -115,16 +121,40 @@ def retrieveFile(self, share_drive_path: str, file_path: str, file) -> (int, int
...
+        path = path if not path or path.endswith(os.sep) else f"{path + os.sep}"
+
+        files_list = []
+        for shared_folder, file_path in SMBConnectionMock.stored_files:
+            shared_folder = shared_folder.replace(os.altsep, os.sep)
+            file_path = file_path.replace(os.altsep, os.sep)
+            file_in_path = to_file_in_path(file_path, path)
+            if (
+                shared_folder == service_name
+                and f"{os.path.dirname(file_path) + os.sep}".startswith(path)
```
Comment (Colin-b) [inline pyndows/testing.py]:
```suggestion
                and f"{os.path.dirname(file_path)}{os.sep}".startswith(path)
```

File: pyndows/testing.py [hunk 27]
```diff
@@ -115,16 +121,40 @@ def retrieveFile(self, share_drive_path: str, file_path: str, file) -> (int, int
...
+                and re.search(pattern, file_in_path[0])
+            ):
+                if (
+                    search | SMB_FILE_ATTRIBUTE_DIRECTORY == search
+                    or not file_in_path[1]
```
Comment (Colin-b) [inline pyndows/testing.py]:
```suggestion
                    or not is_directory
```

File: tests/test_windows.py [hunk 28]
```diff
@@ -187,6 +187,201 @@ def test_file_rename_file_does_not_exist(samba_mock: SMBConnectionMock):
     )
 
 
+def test_get_all_folder_contents(samba_mock: SMBConnectionMock):
+    connection = pyndows.connect(
+        "TestComputer", "127.0.0.1", 80, "TestDomain", "TestUser", "TestPassword"
+    )
+
+    samba_mock.stored_files[("TestShare/", "1")] = "Test Find"
```
Comment (Colin-b) [inline tests/test_windows.py]: Use .update instead of so many single operations

File: tests/test_windows.py [hunk 29]
```diff
@@ -187,6 +187,201 @@ def test_file_rename_file_does_not_exist(samba_mock: SMBConnectionMock):
...
+        connection, "TestShare/", folder_path="A\\i"
+    )
+
+    assert set(shared_folder_contents) == {
+        SharedFileMock(filename="2", isDirectory=False),
+        SharedFileMock(filename="1", isDirectory=False),
+    }
+
+
+def test_get_all_folder_contents_empty_folder(samba_mock: SMBConnectionMock):
```
Comment (Colin-b) [inline tests/test_windows.py]: Can you also test for non existing share in stored files ?

Comment (Colin-b) [reply tests/test_windows.py]: Can you also add test for non existing stored files ?

File: pyndows/testing.py [hunk 30]
```diff
@@ -115,16 +121,36 @@ def retrieveFile(self, share_drive_path: str, file_path: str, file) -> (int, int
...
-        ]
+        pattern = pattern.replace("*", ".*")
+        service_name = service_name.replace(os.altsep, os.sep)
+        path = path.replace(os.altsep, os.sep)
+        path = path if not path or path.endswith(os.sep) else f"{path}{os.sep}"
+
+        def to_file_in_path(file_path: str, path: str) -> (str, bool):
+            file_path = file_path[len(path) :]
+            is_directory = os.sep in file_path
+            if is_directory:
```
Comment (Colin-b) [inline pyndows/testing.py]:
```suggestion
            return file_path.split(os.sep, maxsplit=1)[0], os.sep in file_path
```

**Pull request merged**